tests/test_interpolation.py

```diff
@@ -419,3 +617,381 @@ def test_resolve_key_and_root(
     cfg = _ensure_container(cfg)
     node: Container = OmegaConf.select(cfg, node_key)
     assert node._resolve_key_and_root(key) == expected
+
+
+def test_empty_stack() -> None:
+    """
+    Check that an empty stack during ANTLR parsing raises a `GrammarParseError`.
+    """
+    with pytest.raises(GrammarParseError):
+        grammar_parser.parse("ab}", lexer_mode="VALUE_MODE")
+
+
+def _maybe_create(definition: str) -> Any:
+    """
+    Helper function to create config objects for lists and dictionaries.
+    """
+    if isinstance(definition, (list, dict)):
+        return OmegaConf.create(definition)
+    return definition
+
+
+# Config data used to run many interpolation tests. Each 3-element tuple
+# contains the config key, its value , and its expected value after
+# interpolations are resolved (possibly an exception class).
+# If the expected value is the ellipsis ... then it is expected to be the
+# same as the definition (or `OmegaConf.create()` called on it for lists
+# and dictionaries).
+# Order matters! (each entry should only depend on those above)
+TEST_CONFIG_DATA: List[Tuple[str, Any, Any]] = [
```

Comment:
I was expecting to see some parser specific unit tests (testing the parser directly without going through the higher layers).
Did I miss it?

Reply:
No, I didn't realize it was possible until later, at which point I had already written all tests this way.
Do you want me to refactor them?

Reply:
I missed this one.
Yes -  I think having lower level tests will make the tests more robust and easier to work with.
Once you have lower level tests, this entire function can probably be removed.

Reply:
I refactored these tests in b520815

Essentially I split the big list into 3:
* Stuff that can be parsed without interpolations (to test the basic constructs of the grammar, like int, float, str, dict, etc.). Those can be parsed without using any config object.
* Expressions containing interpolations for the parser rule `singleElement` (this is the biggest chunk)
* Finally, expressions for the highest-level parser rule `configValue`

In the end it still looks pretty similar overall, but at least it's faster and avoids using dummy resolvers everywhere just to test the correct parser rule. Let me know if that looks ok to you.

NB: in a first attempt I tried to go even lower in the rules (e.g., using the `dictContainer` parser rule for dicts), but the problem is that without an `EOF` in the parser rule, ANTLR fails to detect some errors (ex: `{0: 1}}` is not supposed to be a valid dict). That's why I stuck to the `singleElement` and `configValue` rules (that contain the `EOF`).